patterns/strategy.md

```diff
@@ -0,0 +1,217 @@
+    fn run(&self, report: &Report) {
+      print!("[");
+      for (key, val) in report.keys.iter().zip(report.values.iter()) {
+        print!("{{ \"{}\":\"{}\"}},", key, val);
+      }
+      println!("\u{8}]");
+    }
+  }
+}
+
+// Implementation for Plain Text format
+mod vendor2 {
+  use super::*;
+
+  pub struct Text;
+
+  impl Formatter for Text {
+    fn run(&self, report: &Report) {
+      for (key, val) in report.keys.iter().zip(report.values.iter()) {
+        println!("{} {}", key, val);
+      }
+    }
+  }
+}
+
+
+// Report does not implement Generator trait
+// Instead, it provides high level abstraction
+// - vector of Strings: `keys`
+// - vector of integers: `values`
+// - method: `generate`
+// 
+// In other words, Report does not depend on format implementation,
+// but format implementations depend on Report (abtraction). This is called 
+// Dependency Inversion Principle. 
+
+pub struct Report {
+  pub keys:   Vec<String>,
+  pub values: Vec<i32>,
+  // User must provide an object which implements Generator trait
+  formatter: Box<dyn Formatter>
+}
+
+impl Report {
+  pub fn format(&self) {
+    self.formatter.run(&self);
+  }
+
+  pub fn new(formatter: Box<dyn Formatter>) -> Self {
+    let (keys, values) = Self::data_from_db();
+    Report{ keys, values, formatter }
+  }
+
+  fn data_from_db() -> (Vec<String>, Vec<i32>) {
+    let keys = vec!["one".to_string(), "two".to_string()];
+    let values =  vec![1, 2];
+    (keys, values)
+  }
+}
+
+fn main() {
+  let json_report = Report::new(Box::new(vendor1::Json));
+  let text_report = Report::new(Box::new(vendor2::Text));
+
+  println!("JSON format:");
+  json_report.format();
+  
+  println!("\n");
+
+  println!("Plain text format:");
+  text_report.format();
+}
+
+```
+
+## Advantages
+
+Separation of concerns. In the previous example, Report does not know anything about specific implementations of `Vendor1::Json` and `Vendor2::Text`, whereas the output implementations does not care about how data is preprocessed, stored, and fetched. The only thing they have to know is context and and a specific trait and method to implement, i.e., `Generator` and `run`. 
+
+## Disadvantages
+
+For each strategy there must me implemented at least one module, so number of modules increases with number of strategies. If there are many strategies to choose from then users have to know how strategies differ from one another.
+
+## Discussion
+
+In the previous example all strategies in a single file. Typically in Rust, each strategy should be implemented in a separate module file. However, we don't need to use traits in order to design this pattern in Rust. The following toy example demonstates the idea of the Strategy pattern using Rust closures
+
+```rust
+struct Adder;
+impl Adder{
+  pub fn add<F>(x: u8, y: u8, f: F) -> u8
+      where F: Fn(u8, u8) -> u8 {
+      f(x,y)
+  }
+}
+
+fn main() {
+  let arith_adder = |x, y| x + y;
+  let bool_adder = |x, y| if x==1 || y==1 {1} else {0};
```

Comment:
I believe this needs to be formatted.

Reply:
Do you mean spaces? Or one liner `if x==1...`

Reply:
Both spaces and that.